### ISSUE TITLE
MaxPooling3D problem:  miopen requires pooling dimensions be 2.

### ISSUE THREAD

**Issue by kk325ic** [ROCmSoftwarePlatform/tensorflow-upstream#588]
**Describe the current behavior**

Running a simple 3D CNN on a toy example, I get the following running error: 2019-08-01 14:17:00.451253: F tensorflow/stream_executor/rocm/rocm_dnn.cc:759] miopen requires pooling dimensions be 2.

- Have I written custom code in keras:

x = Conv3D(16, (2, 2, 2), activation='relu', padding='same')(inputs) 
x = MaxPooling3D((2, 2, 2), padding='same')(x) 
x = Dropout(0.25)(x)

**System information**
- My system is Ubuntu 18.04 with kernel 4.18.0-25-generic, ROCM 2.6-22 and Tensorflow-rocm 1.14 bare metal
- Python version: 3.6: 
- ROCm/MIOpen version: 2.6-22 
- GPU model and memory: AMD Vega 56, 8G

**Comment by sunway513** [comment]
@jerryyin , can you take a look at this issue?
@kk325ic , it would be helpful if you can provide the steps/scripts to reproduce your issue.

**Comment by jerryyin** [comment]
@sunway513 We don't support `3d pooling` for now, and only support `3d convolution`. I don't know what `miopen` plan is on implementing `3d pooling`. The complaint on pooling dimension is right.

**Comment by sunway513** [comment]
@daniellowell , can you update on the MIOpen plan for `3d pooling`?

**Comment by kk325ic** [comment]
@jerryyin Could you please elaborate your suggested GPU work around? My understanding is that you suggested to reshape the 3D image to 2D, then 2D pooling on it. It seems your suggestion missing the pooling on the first dimension. 

Let us say that I have a 3D images of shape (nZ, nX, nY, channels). After flatten to 2D image, I get (nBatch, nX * nZ, nY, channels).  2D pooling will make the shape to be (nX*nZ / 2, nY/2, channels), which is not equivalent to (nZ/2, nX/2, nY/2). 

An extra 1D pooling might be needed. Please let me know if I missed anything.

**Comment by kk325ic** [comment]
@jerryin , your suggestion of 2D pooling workaround sounds interesting. Let us say that I have the image shape of (nx, ny, nz, channel), then it can reshaped to (ny*nx, nz, channel). Are you suggested to run pooling2D on it? 

If so, it seems need a 1D pooling to make it equivalent to 3D pooling.

**Comment by jerryyin** [comment]
@kk325ic On a second thought, I think doing 2D pooling on each axis is a mistake. Because each pooling on the single axis is only picking local results in the single dimension. Whereas 3D pooling works on "global" 3 dimensions. The correct way to do is instead is to concatenate from (x, y, z, channel) to (xyz, channel), then apply 1D pooling on size of y*z (Any other size will make pooling based on local results thus incorrect). When this is done, reshape it back to (x, y, z, Channel). The downside of this approach is that you cannot have flexible choices on pooling size. Considering this, doing pooling3D on CPU is probably a more convenient approach.

**Comment by kk325ic** [comment]
> @kk325ic On a second thought, I think doing 2D pooling on each axis is a mistake. Because each pooling on the single axis is only picking local results in the single dimension. Whereas 3D pooling works on "global" 3 dimensions. The correct way to do is instead is to concatenate from (x, y, z, channel) to (xyz, channel), then apply 1D pooling on size of y*z (Any other size will make pooling based on local results thus incorrect). When this is done, reshape it back to (x, y, z, Channel). The downside of this approach is that you cannot have flexible choices on pooling size. Considering this, doing pooling3D on CPU is probably a more convenient approach.

Yes, you were right. The GPU work around need to be  reshape to 1D image and reshape back after Pooling1D. 
Is switching between CPU implementation and GPU fast enough? Any suggestion on how to switch MaxPooling3D to CPU?

**Comment by jerryyin** [comment]
@kk325ic Try `with tf.device` to create a device context for GPU. Details refer to [here](https://www.tensorflow.org/guide/using_gpu).

**Comment by kk325ic** [comment]
Thanks for the helps @jerryyin

For the benefits of those who needs MaxPooling3D before MIOpen adds support to it, I implemented a version of workaround by using MaxPooling2D twice. This implementation doesn't support strides not equal to pool_size.

The code is in [here](https://github.com/kk325ic/keras-rocm-workaround/blob/master/mykeras.py).

**Comment by jerryyin** [comment]
MaxPooling3D is supported API now. Closing the issue as solved. Feel free to comment/re-open if this does not solve your question.
